src/iop/useless.c

```diff
@@ -87,18 +113,57 @@ int default_colorspace(dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe, dt_dev_p
   return iop_cs_Lab;
 }
 
+// whenever new fields are added to (or removed from) dt_iop_..._params_t or when their meaning
+// changes, a translation from the old to the new version is added here.
+// A verbatim copy of the old struct definition should be copied into the routine with a _v?_t ending.
+// Since this will get very little future testing (because few developers still have very
+// old versions lying around) existing code should be changed as little as possible, if at all.
+//
+// upgrading from an older version than the previous one should always go through all in between versions
+// (unless there was a bug) so that the end result will always be the same.
```

Comment:
An upgrade path must be provided between any previous version of the module parameters and the current version. For example if the current version is 5 but the old edit is on version 2, the upgrade path must take the parameters cleanly from v2->v3->v4->v5

Reply:
OK, so this bit is somewhat tentative because the full functionality is not yet there in dt_dev_read_history_ext. So _at the moment_ legacy_params needs to provide 1->5, 2->5, 3->5 and 4->5. This leads to the need to amend earlier version upgrades when a new one is added. Some modules already internally implement 1->5 by calling 1->4 and 4->5; this logic should be implemented centrally (which, I think aurelien pointed out already happens for presets or something). In future, once dt_dev_read_history_ext is fixed, only 1->2, 2->3, 3->4 etc need to be provided. At that point it is "guaranteed" to not make a difference if a file is upgraded in one step from 1->5 or, say, from 1->3 and 3->5.
My version doesn't clearly explain this at all. Your example is clear, but the first sentence...
But anyway, this has to be more clearly marked as not operational yet. On my todo list...

Reply:
Ok I didn't realise that (legacy params is one of those bits I'm hazy on).